File: src/main/java/wraith/fabricaeexnihilo/util/ItemUtils.java
package wraith.fabricaeexnihilo.util;

import net.minecraft.block.Block;
import net.minecraft.item.Item;
import net.minecraft.item.ItemConvertible;
import net.minecraft.item.ItemStack;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

import static wraith.fabricaeexnihilo.FabricaeExNihilo.id;

public final class ItemUtils {
    public static ItemStack asStack(ItemConvertible itemConvertible, int amount) {
        return new ItemStack(itemConvertible.asItem(), amount);
    }

    public static ItemStack asStack(ItemConvertible itemConvertible) {
        return new ItemStack(itemConvertible.asItem());
    }

    public static Block getExNihiloBlock(String str) {
        return Registry.BLOCK.get(id(str));
    }

    public static Item getExNihiloItem(String str) {
        return getItem(id(str));
    }

    public static ItemStack getExNihiloItemStack(String str) {
        return getItemStack(id(str));
    }

    public static Item getItem(Identifier identifier) {
        return Registry.ITEM.get(identifier);
    }

    public static ItemStack getItemStack(Identifier identifier) {
        return asStack(getItem(identifier));
    }

    public static ItemStack ofSize(ItemStack stack, int count) {
        var newStack = stack.copy();
        newStack.setCount(count);
        return newStack;
    }
}

File: src/main/kotlin/exnihilofabrico/util/ItemUtils.kt
package exnihilofabrico.util

import exnihilofabrico.id
import me.shedaniel.rei.api.EntryStack
import net.minecraft.entity.ItemEntity
import net.minecraft.item.ItemConvertible
import net.minecraft.item.ItemStack
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import net.minecraft.util.registry.Registry
import net.minecraft.world.World

fun ItemStack.ofSize(num: Int = 1):ItemStack {
    if(num <= 0) return ItemStack.EMPTY
    val stack = this.copy()
    stack.count = num
    return stack
}

fun ItemConvertible.asStack(num: Int = 1) = ItemStack(this.asItem(), num)
fun ItemConvertible.asREIEntry() = EntryStack.create(this.asItem())
fun ItemStack.asREIEntry() = EntryStack.create(this)

fun getItemStack(identifier: Identifier) = Registry.ITEM[identifier].asStack()
fun getExNihiloItemStack(str: String) = getItemStack(id(str))
fun getExNihiloBlock(str: String) = Registry.BLOCK[id(str)]
fun getExNihiloItem(str: String) = Registry.ITEM[id(str)]

fun ItemStack.asEntity(world: World, x: Double, y: Double, z: Double) = ItemEntity(world, x, y, z, this)

fun ItemStack.asEntity(world: World, pos: BlockPos) = this.asEntity(world, pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())

fun ItemStack.spawnAsEntity(world: World, pos: BlockPos) {
    world.spawnEntity(this.asEntity(world, pos))
}

fun World.spawnStack(pos: BlockPos, stack: ItemStack) {
    this.spawnEntity(stack.asEntity(this, pos))
}